Move DockerHub deployment to GitHub Actions

tiemonl opened the pull request:
This closes #14 

Changes:
- Used a matrix to remove the boiler plate code for each release.
- Automated the docker hub release.
  - Currently only pushes linux/amd64 os/arch. See [here](https://hub.docker.com/r/tiemonl/cloudflare-ddns/tags) for example. I'm not quite sure how to push all os/archs. Let me know if you want me to look into it.
  - You will have to add GitHub Actions Secrets `DOCKER_USERNAME` and `DOCKER_PASSWORD`. The password secret would be an access token.
- This could also get rid of [.travis.yml](https://github.com/hugomd/cloudflare-ddns/blob/master/.travis.yml), but I left it just in case.

If you have any questions, please let me know.

Commit:
Use matrix for releases

Commit:
Add docker release

Commit:
Changed Docker Hub repository for pull request

hugomd commented:
Thanks for raising this @tiemonl !

I can configure a `DOCKER_USERNAME` and `DOCKER_PASSWORD` for the main pipeline 👍🏻 

> Currently only pushes linux/amd64 os/arch.

That's totally fine.

> This could also get rid of .travis.yml, but I left it just in case.

It'd be great if you could remove this!

hugomd commented on .github/workflows/release.yml:
```diff
@@ -1,69 +1,32 @@
+          repository: hugomd/cloudflare-ddns
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+          tags: latest
```
This is a really weird requirement, but I wonder if we could avoid pushing to `latest` altogether, and only push for the release tag?

We made some breaking changes in [2.0.0](https://github.com/hugomd/cloudflare-ddns/releases/tag/2.0.0), so I've pinned the `latest` Docker tag to 1.1.1, so we avoid breaking the image for those currently using it. Although, we do display a warning advising migrating to the latest release..

tiemonl replied:
I removed the latest tag. It will only create a tag based on your release tag now. You can see that by comparing the [releases](https://github.com/tiemonl/cloudflare-ddns/releases) on my branch to the releases in [Docker Hub](https://hub.docker.com/r/tiemonl/cloudflare-ddns/tags).

Commit:
remove latest tag

Commit:
Add created release type for a single pipeline.

Commit:
Revert back to hugomd for docker hub for MR.

tiemonl commented:
While fixing the latest tag, I noticed that the pipelines ran three times. I implemented this fix:
```
on: 
  release:
    types: [created]
```
The pipeline will only run once when you create the release. There's different types, [see here](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#release). I was conflicted between released and created, but decided with created as I feel like it will encompass more releases. If you find another type to be more suitable, I can change it. You can see it in action [here](https://github.com/tiemonl/cloudflare-ddns/actions).

Commit:
Remove .travis.yml

hugomd commented:
This looks great, thanks @tiemonl ! `created` should be totally fine for now 👍🏻 

The only thing I've noticed is that it seems to build the docker image for each os/arch and push it to the same tag. I wonder if Docker image should be built and pushed in its own pipeline? Unless there's another way to separate the two?

Commit:
Add job to separate docker hub release.

This will avoid uploading the same tag six times for each binary in docker hub.

tiemonl commented:
> This looks great, thanks @tiemonl ! `created` should be totally fine for now 👍🏻
> 
> The only thing I've noticed is that it seems to build the docker image for each os/arch and push it to the same tag. I wonder if Docker image should be built and pushed in its own pipeline? Unless there's another way to separate the two?

Good catch, I didn't think of that. I separated the jobs out to build first and then release it to Docker Hub. The release step is dependent on the compiling step. If any binary fails, for instance release-windows-amd64, it will not upload to Docker. Let me know what you think.

hugomd commented:
@tiemonl That sounds great! It sounds like you're planning to separate the steps, but to push the binaries first? That sounds perfect!

tiemonl commented:
@hugomd yep! So now it runs `ngs/go-release.action@v1.0.1` which compiles it and appends it to the release for each binary in the `compile` step, then if none of those fail, it will run the `release` step and push it to Docker Hub. That way it only pushes to Docker once. That was achieved with `needs: [compile]`. If the changes are good, it should be good to merge now.

hugomd commented:
That's really neat, thanks for the contributions @tiemonl !

tiemonl commented:
No problem, thank you @hugomd for Dockerising this. I have it running in all my servers now, and it's much easier to set up.

hugomd commented:
My pleasure, please do let me know if there's any features you'd like added! 👍🏻
